src/main/java/net/snowflake/client/jdbc/cloud/storage/SnowflakeGCSClient.java

```diff
@@ -897,37 +887,35 @@ public void handleStorageException(
           // ignore
         }
 
-        if (se.getCode() == 403 && session != null && command != null) {
-          // A 403 indicates that the access token has expired,
+        if (se.getCode() == 401 && session != null && command != null) {
```

Comment:
Are you sure that GCS never returns a 403 on an expired token/URL?

Reply:
Yes, I am pretty confident in that. When using presigned url, the error code would be 400, and access token is 401 is expired. For access token, this follows Oauth, which says 401 error code for expiration, and in Google's sdk GoogleCredential code, it also refresh for 401 error.

I did some real time testing and confirmed it also.

Reply:
I did a little digging to figure out how we ended up with this code, and it seems that Azure returns a 403 on expired SAS tokens. It seems that GCS returns a different error code on an expired token. From this SO issue, it seems like that error code may be 400: https://stackoverflow.com/questions/55059892/google-storage-signed-url-expiration-results-in-400-error-with-no-cors-headers
And your testing shows that it's a 401. My advice is that we test against 400, 403, and 401 to cover our bases.

Reply:
@sfc-gh-ppaulus I added error throw for presigned url failure.

Reply:
400 is for pre-signed URL only, that code path is only enter when using access token so I don't think it is an issue. So adding 400 and 403 would pretty much be useless. 

Also note that the refresh token being called in there only refresh the access token, presigned url remain unchange, so there is no reason to check for 400 still.

Reply:
https://cloud.google.com/storage/docs/xml-api/reference-status#401%E2%80%94unauthorized

Reply:
What makes this codepath only for tokens?

Reply:
In the old code, the function return after response receive after execute presigned url (regardless whether it fail or success, as long as exception is not thrown). Also this code path is when the exception being pass in is StorageException  from Google Storage API, which use access token.

Reply:
I don't quite understand why this code wouldn't get hit when using a presigned url if an exception is thrown, which is what I would expect the SDK to do if the request fails.

Reply:
Presigned url execution  does not use sdk, it is purely http request/response, which also don't throw error.

Reply:
Presigned url implementation do not throw exception. And even if it throws, it would not throw StorageException, since it is not a part of the sdk.

Reply:
Ah, yes, you're right. Ok, can you make a comment in the code that this is only for tokens because presigned url codepaths don't raise storage eexceptions?

Reply:
Added comment.

Reply:
Thanks Khang!